### ISSUE TITLE
Introduce intermediate data structure between parsetree and Format stream

### ISSUE THREAD

**Issue by jberdine** [ocaml-ppx/ocamlformat#406]
The current implementation traverses the parsetree and directly produces the `Format` stream. This works well when the desired nesting structure of formatting boxes coincides with the nesting structure of the parsetree. When the desired formatting involves different nesting of boxes than the parsetree, the code gets very complicated and non-compositional.

An alternative is to introduce an intermediate data structure which is constructed by traversing the parsetree, and then it is traversed to produce the `Format` stream. This would enable much simpler and more flexible differences between nesting of boxes relative to parsetree structure.

Currently all the various "sugar" functions, as well as `Fmt_ast.block`, are partial steps in this direction already.

**Comment by gpetiot** [comment]
I lowered the priority of this one, as it looks very time consuming and not critical for now.

**Comment by jberdine** [comment]
Yes, I don't know when the right time to do this will be. At some point we will have spent more time (adding up small pieces) fighting with the current implementation than it would have taken to implement this. But it's hard to predict.
